In [54]:
#libraries
import pandas as pd
import scipy.stats as st
import numpy as np

In [55]:
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/pokemon.csv")
df

,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,Mega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
4,Charmander,Fire,NaN,39,52,43,60,50,65,1,False
...,...,...,...,...,...,...,...,...,...,...,...
795,Diancie,Rock,Fairy,50,100,150,100,150,50,6,True
796,Mega Diancie,Rock,Fairy,50,160,110,160,110,110,6,True
797,Hoopa Confined,Psychic,Ghost,80,110,60,150,130,70,6,True
798,Hoopa Unbound,Psychic,Dark,80,160,60,170,130,80,6,True


In [56]:
# Step 1: Extract the two groups
dragon_hp = df[df['Type 1'] == 'Dragon']['HP']
grass_hp = df[df['Type 1'] == 'Grass']['HP']

print("Dragon HP - n:", len(dragon_hp), " mean:", dragon_hp.mean(), " std:", dragon_hp.std())
print("Grass HP - n:", len(grass_hp), " mean:", grass_hp.mean(), " std:", grass_hp.std())

Dragon HP - n: 32  mean: 83.3125  std: 23.795414982587474
Grass HP - n: 70  mean: 67.27142857142857  std: 19.516564074957593


In [57]:
alpha = 0.05

In [58]:
# Normality (Shapiro-Wilk, since n < 5000 for both groups)
_, p_dragon_norm = st.shapiro(dragon_hp)
_, p_grass_norm = st.shapiro(grass_hp)
print("Shapiro Dragon p-value:", p_dragon_norm)
print("Shapiro Grass p-value:", p_grass_norm)

# Homogeneity of variance (Levene's test)
_, p_levene = st.levene(dragon_hp, grass_hp)
print("Levene p-value:", p_levene)

Shapiro Dragon p-value: 0.3419890019248819
Shapiro Grass p-value: 0.11214356726520774
Levene p-value: 0.1748473993433107


In [59]:
_, p_value = st.ttest_ind(dragon_hp, grass_hp, equal_var=False, alternative="greater")
print(p_value)

0.0007993609745420599


In [60]:
if p_value < alpha:
    print("The null hypothesis H0 is rejected -> The alternative hypothesis H1 is accepted.")
else:
    print("The null hypothesis H0 is accepted.")

The null hypothesis H0 is rejected -> The alternative hypothesis H1 is accepted.


Conclusion: There is statistically significant evidence that Dragon-type Pokémon have, on average, higher HP than Grass-type Pokémon. The sample means support this too (Dragon mean HP ≈ 83.3 vs Grass mean HP ≈ 67.3), and since the p-value is far below 0.05, this difference is very unlikely to be due to random chance.

In [61]:
legendary = df[df["Legendary"] == True]
non_legendary = df[df["Legendary"] == False]

In [62]:
leg_hp = legendary["HP"]
nonleg_hp = non_legendary["HP"]

# Normality
print(st.shapiro(leg_hp))
print(st.shapiro(nonleg_hp))

# Equal variances
print(st.levene(leg_hp, nonleg_hp))

ShapiroResult(statistic=np.float64(0.948212604837747), pvalue=np.float64(0.008645211266956951))
ShapiroResult(statistic=np.float64(0.8959988676960179), pvalue=np.float64(6.930385498829385e-22))
LeveneResult(statistic=np.float64(0.31326136040404007), pvalue=np.float64(0.5758428070138398))


In [63]:
_, p_value_hp = st.mannwhitneyu(leg_hp, nonleg_hp, alternative="two-sided")
print(p_value_hp)

6.93110816444599e-17


In [64]:
leg_atk = legendary["Attack"]
nonleg_atk = non_legendary["Attack"]

print(st.shapiro(leg_atk))
print(st.shapiro(nonleg_atk))
print(st.levene(leg_atk, nonleg_atk))

ShapiroResult(statistic=np.float64(0.9607938487630754), pvalue=np.float64(0.03771789634848271))
ShapiroResult(statistic=np.float64(0.9808355320365648), pvalue=np.float64(3.1843612080590274e-08))
LeveneResult(statistic=np.float64(4.816137849160818e-05), pvalue=np.float64(0.9944645877795636))


In [65]:
_, p_value_atk = st.mannwhitneyu(leg_atk, nonleg_atk, alternative="two-sided")
print(p_value_atk)

3.4625666210159153e-19


In [66]:
leg_def = legendary["Defense"]
nonleg_def = non_legendary["Defense"]

print(st.shapiro(leg_def))
print(st.shapiro(nonleg_def))
print(st.levene(leg_def, nonleg_def))

ShapiroResult(statistic=np.float64(0.9419734556259307), pvalue=np.float64(0.00430961213067538))
ShapiroResult(statistic=np.float64(0.9267812555700486), pvalue=np.float64(1.8918446076661715e-18))
LeveneResult(statistic=np.float64(0.9887769396854924), pvalue=np.float64(0.3203429011757649))


In [67]:
_, p_value_def = st.mannwhitneyu(leg_def, nonleg_def, alternative="two-sided")
print(p_value_def)

3.390882856434108e-14


In [68]:
leg_spatk = legendary["Sp. Atk"]
nonleg_spatk = non_legendary["Sp. Atk"]

print(st.shapiro(leg_spatk))
print(st.shapiro(nonleg_spatk))
print(st.levene(leg_spatk, nonleg_spatk))

ShapiroResult(statistic=np.float64(0.9850924611860353), pvalue=np.float64(0.6240339399453302))
ShapiroResult(statistic=np.float64(0.9675503038499106), pvalue=np.float64(1.0736763014225767e-11))
LeveneResult(statistic=np.float64(0.8040982204951724), pvalue=np.float64(0.37014131900583735))


In [69]:
_, p_value_spatk = st.mannwhitneyu(leg_spatk, nonleg_spatk, alternative="two-sided")
print(p_value_spatk)

6.287374046693629e-26


In [70]:
leg_spdef = legendary["Sp. Def"]
nonleg_spdef = non_legendary["Sp. Def"]

print(st.shapiro(leg_spdef))
print(st.shapiro(nonleg_spdef))
print(st.levene(leg_spdef, nonleg_spdef))

ShapiroResult(statistic=np.float64(0.9477878830687874), pvalue=np.float64(0.008238834358460644))
ShapiroResult(statistic=np.float64(0.9623620065493651), pvalue=np.float64(8.430313884525228e-13))
LeveneResult(statistic=np.float64(0.08071031493846148), pvalue=np.float64(0.7764108095205825))


In [71]:
_, p_value_spdef = st.mannwhitneyu(leg_spdef, nonleg_spdef, alternative="two-sided")
print(p_value_spdef)

1.2966607518348966e-20


In [72]:
leg_speed = legendary["Speed"]
nonleg_speed = non_legendary["Speed"]

print(st.shapiro(leg_speed))
print(st.shapiro(nonleg_speed))
print(st.levene(leg_speed, nonleg_speed))

ShapiroResult(statistic=np.float64(0.9358058164164608), pvalue=np.float64(0.002216855219004437))
ShapiroResult(statistic=np.float64(0.9835871527438954), pvalue=np.float64(2.4626530460697836e-07))
LeveneResult(statistic=np.float64(9.823256712658159), pvalue=np.float64(0.0017862391470896439))


In [73]:
_, p_value_speed = st.mannwhitneyu(leg_speed, nonleg_speed, alternative="two-sided")
print(p_value_speed)

8.147512738479629e-19


For every single stat, at least one group (usually the Non-Legendary one, since it's much bigger and more varied) failed the normality check, so Mann-Whitney U was the correct test in all 6 cases — never the t-test.

And for every single stat, the p-value came back astronomically small — nowhere close to 0.05, all in the range of e-14 to e-26. Since all six p-values are far below α = 0.05, we reject H0 in every case.

In [74]:
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/california_housing.csv")
df.head()

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value
0,-114.31,34.19,15.0,5612.0,1283.0,1015.0,472.0,1.4936,66900.0
1,-114.47,34.40,19.0,7650.0,1901.0,1129.0,463.0,1.8200,80100.0
2,-114.56,33.69,17.0,720.0,174.0,333.0,117.0,1.6509,85700.0
3,-114.57,33.64,14.0,1501.0,337.0,515.0,226.0,3.1917,73400.0
4,-114.57,33.57,20.0,1454.0,326.0,624.0,262.0,1.9250,65500.0


In [75]:
alpha = 0.05

In [76]:
def euclidean_distance(lat1, lon1, lat2, lon2):
    return ((lat2 - lat1)**2 + (lon2 - lon1)**2) ** 0.5

In [77]:
school_lon, school_lat = -118, 34
hospital_lon, hospital_lat = -122, 37


In [78]:
df["dist_school"] = euclidean_distance(df["latitude"], df["longitude"], school_lat, school_lon)
df["dist_hospital"] = euclidean_distance(df["latitude"], df["longitude"], hospital_lat, hospital_lon)

In [79]:
df["close"] = (df["dist_school"] < 0.50) | (df["dist_hospital"] < 0.50)

In [80]:
close_price = df[df["close"] == True]["median_house_value"]
far_price = df[df["close"] == False]["median_house_value"]

In [81]:
alpha = 0.05

In [82]:
st.kstest(close_price, "norm", args=(close_price.mean(), close_price.std()))

KstestResult(statistic=np.float64(0.11883225900685446), pvalue=np.float64(1.7527077264617067e-84), statistic_location=np.float64(237500.0), statistic_sign=np.int8(1))

In [83]:
st.kstest(far_price, "norm", args=(far_price.mean(), far_price.std()))

KstestResult(statistic=np.float64(0.12112013386816922), pvalue=np.float64(1.742313576944486e-130), statistic_location=np.float64(165600.0), statistic_sign=np.int8(1))

In [84]:
st.levene(close_price, far_price)

LeveneResult(statistic=np.float64(5.34456402848745), pvalue=np.float64(0.02079887481140396))

In [85]:
_, p_value = st.mannwhitneyu(close_price, far_price, alternative="greater")
print(p_value)


0.0


In [86]:
if p_value < alpha:
    print("The null hypothesis H0 is rejected -> The alternative hypothesis H1 is accepted.")
else:
    print("The null hypothesis H0 is accepted.")

The null hypothesis H0 is rejected -> The alternative hypothesis H1 is accepted.


Houses that are close to a school or hospital tend to cost more than houses that are far away — and this difference is real, not just random chance.

We know it's not random chance because the test gave us a p-value that's basically zero, which is way below our 0.05 cutoff. With such a large dataset (over 17,000 houses), we can trust this result.